src/main/java/ladysnake/effective/client/world/WaterfallCloudGenerators.java

```diff
@@ -64,13 +64,14 @@ public static void tick() {
     }
 
     private static void tickParticles(World world) {
-        for (BlockPos pos : particlesToSpawn.keySet()) {
-            if (pos != null) {
-                if (particlesToSpawn.put(pos, particlesToSpawn.getInt(pos) - 1) <= 0) {
-                    particlesToSpawn.removeInt(pos);
+        synchronized (particlesToSpawn) {
+            for (BlockPos pos : particlesToSpawn.keySet()) {
+                if (pos != null) {
+                    particlesToSpawn.put(pos, particlesToSpawn.getInt(pos) - 1);
+                    addWaterfallCloud(world, pos);
                 }
-                addWaterfallCloud(world, pos);
             }
+            particlesToSpawn.object2IntEntrySet().removeIf(blockPosEntry -> blockPosEntry.getIntValue() < 0);
         }
     }
```

Comment:
`synchronized` doesn't do anything because particles aren't ticked on different threads. Additionally, `Map`s include methods to do the removals in a more streamlined manner:

```suggestion
    private static void tickParticles(World world) {
        for (BlockPos pos : particlesToSpawn.keySet()) {
            if (pos != null) {
                particlesToSpawn.computeInt(pos, (blockPos, integer) -> integer - 1);
                addWaterfallCloud(world, pos);
            }
        }
        particlesToSpawn.values().removeIf(integer -> integer < 0);
    }
```

Reply:
It's shorter! You've been very helpful.
And..., I found a bit of mistake.

Reply:
This suggestion is included in 34fdfba .